File: 2023/python/aoc2023/day04.py
import sys
from typing import TextIO
from aoc2023 import Solve


class Solution(Solve):
    def __init__(self, io: TextIO) -> None:
        def match_count(s: str) -> int:
            winning, have = map(str.split, s.split(":")[1].split("|"))
            return len(set(winning) & set(have))

        self.matches = list(map(match_count, io))

    def part1(self) -> int:
        return sum(1 << i >> 1 for i in self.matches)

    def part2(self) -> int:
        cards = [1] * len(self.matches)
        for i, m in enumerate(self.matches):
            for j in range(m):
                cards[i + 1 + j] += cards[i]
        return sum(cards)


if __name__ == "__main__":
    solution = Solution(sys.stdin)
    print(f"Part 1: {solution.part1()}")
    print(f"Part 2: {solution.part2()}")

File: 2023/ocaml/lib/day04.ml
open Base
open Solution

module Solution : Solve = struct
  type t = int list

  let parse input =
    let match_count line =
      let to_int_list s =
        String.split s ~on:' '
        |> List.filter ~f:(Fn.non String.is_empty)
        |> List.map ~f:Int.of_string
      in
      String.lsplit2_exn line ~on:':' |> snd |> String.lsplit2_exn ~on:'|'
      |> fun (winning, have) ->
      let w = winning |> to_int_list in
      have |> to_int_list |> List.count ~f:(List.mem w ~equal)
    in
    In_channel.input_lines input |> List.map ~f:match_count

  let part1 matches =
    List.sum (module Int) matches ~f:(fun x -> (1 lsl x) lsr 1) |> Answer.of_int

  let part2 matches =
    let a = Array.create ~len:(List.length matches) 1 in
    List.iteri matches ~f:(fun i x ->
        List.range (i + 1) (i + 1 + x)
        |> List.iter ~f:(fun j -> a.(j) <- a.(j) + a.(i)));
    Array.fold a ~init:0 ~f:( + ) |> Answer.of_int
end
